File: notes_app/lib/screens/notes_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:appwrite/models.dart';
import '../components/note_item.dart';
import '../components/note_input_dialog.dart';
import '../providers/auth_provider.dart';
import '../services/note_service.dart';
import '../widgets/logout_button.dart';

class NotesScreen extends StatefulWidget {
  const NotesScreen({super.key});

  @override
  // ignore: library_private_types_in_public_api
  _NotesScreenState createState() => _NotesScreenState();
}

class _NotesScreenState extends State<NotesScreen> {
  // Note service
  final NoteService _noteService = NoteService();
  
  // Notes list from Appwrite
  List<Document> notes = [];
  
  // Loading state
  bool isLoading = true;

  // Controllers and state variables
  TextEditingController noteController = TextEditingController();
  String? editingNoteId;

  @override
  void initState() {
    super.initState();
    _fetchNotes();
  }

  // Fetch notes from Appwrite
  Future<void> _fetchNotes() async {
    final authProvider = Provider.of<AuthProvider>(context, listen: false);
    if (authProvider.user == null) {
      setState(() {
        isLoading = false;
      });
      return;
    }

    setState(() {
      isLoading = true;
    });

    try {
      // Get user ID
      String userId = authProvider.user.$id;
      
      // Fetch notes for this user
      final fetchedNotes = await _noteService.getNotes(userId: userId);
      setState(() {
        notes = fetchedNotes;
        isLoading = false;
      });
    } catch (e) {
      print('Failed to fetch notes: $e');
      setState(() {
        isLoading = false;
      });
    }
  }

  // Function to save a new or updated note
  Future<void> saveNote() async {
    if (noteController.text.trim().isEmpty) return;

    final authProvider = Provider.of<AuthProvider>(context, listen: false);
    if (authProvider.user == null) return;

    try {
      String userId = authProvider.user.$id;
      
      if (editingNoteId != null) {
        // Update existing note
        final updatedNote = await _noteService.updateNote(
          editingNoteId!,
          {
            'content': noteController.text,
            'userId': userId,
          },
        );
        
        setState(() {
          int index = notes.indexWhere((note) => note.$id == editingNoteId);
          if (index != -1) {
            notes[index] = updatedNote;
          }
          editingNoteId = null;
        });
      } else {
        // Add new note
        final newNote = await _noteService.createNote({
          'content': noteController.text,
          'userId': userId,
        });
        
        setState(() {
          notes.insert(0, newNote);
        });
      }
      
      noteController.clear();
      Navigator.of(context).pop(); // Close the dialog
    } catch (e) {
      print('Failed to save note: $e');
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(content: Text('Failed to save note: $e')),
      );
    }
  }

  // Function to delete a note
  Future<void> deleteNote(String id) async {
    try {
      await _noteService.deleteNote(id);
      setState(() {
        notes.removeWhere((note) => note.$id == id);
      });
    } catch (e) {
      print('Failed to delete note: $e');
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(content: Text('Failed to delete note: $e')),
      );
    }
  }

  // Function to open edit mode
  void editNote(Map<String, dynamic> note) {
    editingNoteId = note['\$id'];
    noteController.text = note['content'] ?? '';
    showNoteDialog();
  }

  // Show dialog for adding/editing notes
  void showNoteDialog() {
    showDialog(
      context: context,
      builder: (context) => NoteInputDialog(
        controller: noteController,
        isEditing: editingNoteId != null,
        onSave: saveNote,
      ),
    );
  }

  // Convert Document to Map for compatibility with NoteItem widget
  Map<String, dynamic> _documentToMap(Document doc) {
    return {
      '\$id': doc.$id,
      'id': doc.$id,
      'content': doc.data['content'] ?? '',
      'createdAt': doc.data['createdAt'] ?? DateTime.now().toIso8601String(),
      'updatedAt': doc.data['updatedAt'],
    };
  }

  @override
  Widget build(BuildContext context) {
    final authProvider = Provider.of<AuthProvider>(context);
    
    return Scaffold(
      appBar: AppBar(
        title: const Text('My Notes'),
        backgroundColor: Colors.blue,
        foregroundColor: Colors.white,
        actions: const [
          LogoutButton(),
        ],
      ),
      body: isLoading
          ? const Center(child: CircularProgressIndicator())
          : notes.isEmpty
              ? Center(
                  child: Column(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: [
                      const Text(
                        "You don't have any notes yet.",
                        style: TextStyle(
                          fontSize: 18,
                          fontWeight: FontWeight.bold,
                        ),
                      ),
                      const SizedBox(height: 10),
                      Text(
                        "Tap the + button to create your first note!",
                        style: TextStyle(
                          fontSize: 16,
                          color: Colors.grey[600],
                        ),
                        textAlign: TextAlign.center,
                      ),
                    ],
                  ),
                )
              : RefreshIndicator(
                  onRefresh: _fetchNotes,
                  child: ListView.builder(
                    padding: const EdgeInsets.all(15),
                    itemCount: notes.length,
                    itemBuilder: (context, index) {
                      return NoteItem(
                        note: _documentToMap(notes[index]),
                        onEdit: editNote,
                        onDelete: (id) => deleteNote(id),
                      );
                    },
                  ),
                ),
      floatingActionButton: FloatingActionButton(
        backgroundColor: Colors.blue,
        child: const Icon(Icons.add, color: Colors.white),
        onPressed: () {
          editingNoteId = null;
          noteController.clear();
          showNoteDialog();
        },
      ),
    );
  }

  @override
  void dispose() {
    // Clean up the controller when the widget is disposed
    noteController.dispose();
    super.dispose();
  }
}

File: notes_app_react/screens/NotesScreen.js
import React, { useState, useEffect } from "react";
import {
  View,
  Text,
  StyleSheet,
  FlatList,
  TouchableOpacity,
  ActivityIndicator,
} from "react-native";
import NoteItem from "../components/NoteItem";
import NoteInput from "../components/NoteInput";
import { useAuth } from "../contexts/AuthContext";
import { getNotes, createNote, deleteNote as deleteNoteService, updateNote as updateNoteService } from "../services/note-service";

export default function NotesScreen({ navigation }) {
  const [notes, setNotes] = useState([]);
  const [modalVisible, setModalVisible] = useState(false);
  const [noteText, setNoteText] = useState("");
  const [editingNote, setEditingNote] = useState(null);
  const [isLoading, setIsLoading] = useState(true);
  const { user, isAuthenticated, loading } = useAuth();

  // Redirect unauthenticated users to Auth screen
  useEffect(() => {
    if (!loading && !isAuthenticated) {
      navigation.replace("Auth");
    }
  }, [isAuthenticated, loading, navigation]);

  // Fetch notes when user is available
  const fetchNotes = async () => {
    if (!user) return;
    
    setIsLoading(true);
    try {
      // Pass the user ID when fetching notes
      const fetchedNotes = await getNotes(user.$id);
      setNotes(fetchedNotes);
    } catch (error) {
      console.error("Failed to fetch notes:", error);
    } finally {
      setIsLoading(false);
    }
  };

  useEffect(() => {
    if (user) {
      fetchNotes();
    }
  }, [user]);

  // Function to add or update a note
  const saveNote = async () => {
    if (noteText.trim() === "") return;

    try {
      if (editingNote) {
        // Update existing note
        const updatedNote = await updateNoteService(editingNote.id, {
          content: noteText,
        });
        setNotes(notes.map((note) => (note.id === editingNote.id ? updatedNote : note)));
        setEditingNote(null);
      } else {
        // Add new note with user ID
        const newNote = await createNote({
          content: noteText,
          userId: user.$id,
        });
        setNotes([newNote, ...notes]);
      }

      setNoteText("");
      setModalVisible(false);
    } catch (error) {
      console.error("Failed to save note:", error);
    }
  };

  // Function to delete a note
  const handleDeleteNote = async (id) => {
    try {
      await deleteNoteService(id);
      setNotes(notes.filter((note) => note.id !== id));
    } catch (error) {
      console.error("Failed to delete note:", error);
    }
  };

  // Function to open edit mode
  const editNote = (note) => {
    setEditingNote(note);
    setNoteText(note.content);
    setModalVisible(true);
  };

  // Function to close the modal
  const closeModal = () => {
    setModalVisible(false);
    setNoteText("");
    setEditingNote(null);
  };

  const renderEmptyComponent = () => {
    return (
      <View style={styles.emptyContainer}>
        <Text style={styles.emptyText}>You don't have any notes yet.</Text>
        <Text style={styles.emptySubtext}>
          Tap the + button to create your first note!
        </Text>
      </View>
    );
  };

  if (loading) {
    return (
      <View style={styles.container}>
        <ActivityIndicator size="large" color="#3498db" />
      </View>
    );
  }

  return (
    <View style={styles.container}>
      <View style={styles.header}>
        <Text style={styles.headerTitle}>My Notes</Text>
        <TouchableOpacity
          style={styles.addButton}
          onPress={() => setModalVisible(true)}
        >
          <Text style={styles.addButtonText}>+</Text>
        </TouchableOpacity>
      </View>

      {isLoading ? (
        <View style={styles.loadingContainer}>
          <ActivityIndicator size="large" color="#3498db" />
        </View>
      ) : (
        <FlatList
          data={notes}
          renderItem={({ item }) => (
            <NoteItem note={item} onEdit={editNote} onDelete={handleDeleteNote} />
          )}
          keyExtractor={(item) => item.id}
          contentContainerStyle={notes.length === 0 ? { flex: 1 } : styles.notesList}
          ListEmptyComponent={renderEmptyComponent()}
        />
      )}

      <NoteInput
        visible={modalVisible}
        onClose={closeModal}
        onSave={saveNote}
        noteText={noteText}
        setNoteText={setNoteText}
        isEditing={!!editingNote}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#f5f5f5",
  },
  header: {
    height: 100,
    backgroundColor: "#3498db",
    flexDirection: "row",
    justifyContent: "space-between",
    alignItems: "flex-end",
    paddingBottom: 15,
    paddingHorizontal: 20,
  },
  headerTitle: {
    color: "white",
    fontSize: 24,
    fontWeight: "bold",
  },
  addButton: {
    width: 36,
    height: 36,
    borderRadius: 18,
    backgroundColor: "white",
    justifyContent: "center",
    alignItems: "center",
  },
  addButtonText: {
    fontSize: 24,
    color: "#3498db",
    fontWeight: "bold",
    marginTop: -2,
  },
  notesList: {
    padding: 15,
  },
  emptyContainer: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    padding: 20,
  },
  emptyText: {
    fontSize: 18,
    fontWeight: "bold",
    marginBottom: 10,
    textAlign: "center",
  },
  emptySubtext: {
    fontSize: 16,
    color: "#666",
    textAlign: "center",
  },
  loadingContainer: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
  },
});